drainer/k8s_utils.py

```diff
@@ -28,19 +30,37 @@ def cordon_node(api, node_name):
 
 def remove_all_pods(api, node_name, poll=5):
     """Removes all Kubernetes pods from the specified node."""
-    field_selector = 'spec.nodeName=' + node_name
-    pods = api.list_pod_for_all_namespaces(watch=False, field_selector=field_selector)
+    pods = get_evictable_pods(api, node_name)
+
+    logger.debug('Number of pods to delete: ' + str(len(pods)))
+
+    evict_until_completed(api, pods, poll)
+    wait_until_empty(api, node_name, poll)
 
-    logger.debug('Number of pods to delete: ' + str(len(pods.items)))
 
-    try_until_completed(api, evict_pods, pods, poll)
-    try_until_completed(api, get_pending, pods, poll)
+def pod_is_evictable(pod):
+    if pod.metadata.annotations is not None and pod.metadata.annotations.get(MIRROR_POD_ANNOTATION_KEY):
+        logger.info("Skipping mirror pod %s/%s", pod.metadata.namespace, pod.metadata.name)
+        return False
+    if pod.metadata.owner_references is None:
+        return True
+    for ref in pod.metadata.owner_references:
+        if ref.controller is not None and ref.controller:
+            if ref.kind == CONTROLLER_KIND_DAEMON_SET:
+                logger.info("Skipping DaemonSet %s/%s", pod.metadata.namespace, pod.metadata.name)
+                return False
+    return True
 
 
-def try_until_completed(api, action, pods, poll):
-    pending = pods.items
-    while len(pending) > 0:
-        pending = action(api, pending)
+def get_evictable_pods(api, node_name):
+    field_selector = 'spec.nodeName=' + node_name
+    pods = api.list_pod_for_all_namespaces(watch=False, field_selector=field_selector, include_uninitialized=True)
+    return list(filter(pod_is_evictable, pods.items))
```

Comment:
To avoid iterating over the list twice by having to cast, let's just use a list comprehension.